### PR TITLE
Migration: 'closestTo' functions to TS

### LINKED ISSUE #1915: `closestTo` TS typings are incorrect
If you give the `closestTo` function an empty list of dates, it returns `undefined`. This is not represented in the TS typings.

### PR DESCRIPTION
**Pull request by V-Gutierrez**
- Adapt implementations to TS
- Change tests and documentation to match functions current behavior ( extends #1933 closes #1915   ) 
- Remove deprecated asserts
- Update variables
- Closes date-fns/date-fns-typescriptify#1 and closes date-fns/date-fns-typescriptify#2

### CODE REVIEW

File: src/closestTo/test.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+/* eslint-env mocha */
+
+import assert from 'assert'
+import closestTo from '.'
+
+describe('closestTo', function() {
+  it('returns the date from the given array closest to the given date', function() {
+    const date = new Date(2014, 6 /* Jul */, 2)
+    const result = closestTo(date, [
+      new Date(2015, 7 /* Aug */, 31),
```
Comment (hjJunior) [inline src/closestTo/test.ts:11]: WDUT about using the initialization by formatted string date? 

eg.

```suggestion
      new Date("2015-08-31 00:00:00"),
```

Comment (V-Gutierrez) [reply src/closestTo/test.ts:11]: That is a good idea!

But I'm trying to keep these consistent with the way they're used in the entire library tests, at least in this first round of migrations. 

Also, I think the way it is, leads to a more readable, more maintainable, and easier to contribute code. 

Do you agree with my point? :D

Comment (kossnocorp) [reply src/closestTo/test.ts:11]: Let's keep it as is. Besides `"2015-08-31 00:00:00"` is an invalid ISO string.

**Comment by maximtop** [comment]
Hi! Could you please rebase you pr and mark that it is closing following issues
https://github.com/date-fns/date-fns-typescriptify/issues/1
https://github.com/date-fns/date-fns-typescriptify/issues/2

**Comment by V-Gutierrez** [comment]
> Hi! Could you please rebase you pr and mark that it is closing following issues
> [date-fns/date-fns-typescriptify#1](https://github.com/date-fns/date-fns-typescriptify/issues/1)
> [date-fns/date-fns-typescriptify#2](https://github.com/date-fns/date-fns-typescriptify/issues/2)


Good to go

**Comment by maximtop** [comment]
Something went wrong, there shouldn't be 133 commits in you commit. 
May be it's worth to close this pr, fetch changes from master and open the new one only with your commits. Otherwise it is impossible to review this pr.

**Comment by V-Gutierrez** [comment]
Done

**Review by tan75 — APPROVED** [review]
🙏. thank you!

**Review by tan75 — APPROVED** [review]

**Pull request merged**